chore: add support to build images which generate coverage data

blaisedias opened the pull request:
- coverage data is generated in /var/local/mayastor-coverage
- update yamls file with necessary runes
- nix changes to build appropriate images
    -cov for coverage
    -dev for debug
- coverage builds require rust nightly toolset
    - "-Z instrument-coverage"
- update release.sh to build coverage images

- add signal handlers for SIGTERM and SIGINT to csi-controller   
    On receiving SIGTERM or SIGINT, shutdown in an orderly  manner to generate coverage data.

Commit:
chore: add support to build images which generate coverage data

- coverage data is generated in /var/local/mayastor-coverage
- update yamls file with necessary runes
- nix changes to build appropriate images
    -cov for coverage
    -dev for debug
- coverage builds require rust nightly toolset
    - "-Z instrument-coverage"
- update release.sh to build coverage images

blaisedias commented on nix/pkgs/images/default.nix:
```diff
@@ -10,54 +10,78 @@
     tag = control-plane.version;
     created = "now";
-    name = "mayadata/mcp-${name}";
+    name = "mayadata/mcp-${name}${variant}";
```
@tiagolobocastro I couldn't get your last fix,  to work, however on reflection we shouldn't default suffixes. So what was added in https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/fe5d9a0d03035df7fa0e55daf5c904d6216413a4 should be okay

Commit:
chore: add signal handlers for SIGTERM and SIGINT

On receiving SIGTERM or SIGINT, shutdown in an orderly
manner to generate coverage data.

tiagolobocastro commented on chart/templates/core-agents-deployment.yaml:
```diff
@@ -39,4 +39,14 @@ spec:
-                  fieldPath: metadata.namespace
\ No newline at end of file
+                  fieldPath: metadata.namespace
+            - name: LLVM_PROFILE_FILE
```
It's ignored but it means we're always creating folders in the host systems because of the hostPath, and in that past that has caused issues so IMHO we shouldn't add that to the yaml's by default.

gila replied:
This should be part of the "thing" that does the coverage.

blaisedias replied:
Fixed.
https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/c8df674795035432102a90e95b891f9627908aaa

tiagolobocastro commented on control-plane/csi-controller/src/main.rs:
```diff
@@ -97,10 +97,27 @@ pub async fn main() -> Result<(), String> {
+                    }
+            }
+        }
+        _ => Err("failed to add signal handlers".to_string()),
```
If the signal handlers are a requisite, then you can do eg .expect("Failed to register SIGTERM handler") on line 100 and then we'll know exactly which one failed.

blaisedias replied:
Done https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/08896f39ff98e16a05b6014492cc5a70398050d4

tiagolobocastro commented on nix/overlay.nix:
```diff
@@ -3,4 +3,5 @@ self: super: {
   control-plane = super.callPackage ./pkgs/control-plane { };
   utils = super.callPackage ./pkgs/utils { };
   openapi-generator = super.callPackage ./pkgs/openapi-generator { };
+  control-plane-cov = (super.callPackage ./pkgs/control-plane { }).cov;
```
can be deleted as you're not using this AFAICT

blaisedias replied:
Done.

blaisedias commented on control-plane/csi-controller/src/main.rs:
```diff
@@ -97,10 +97,27 @@ pub async fn main() -> Result<(), String> {
+                _srv =  server::CsiServer::run(args.value_of("socket")
+                    .expect("CSI socket must be specfied")
+                    .to_string()) => {
+                    Ok(())
```
Done.

Commit:
fix: remove unnecessary directive

blaisedias commented:
bors try

gila commented on chart/templates/csi-deployment.yaml:
```diff
@@ -61,10 +61,17 @@ spec:
           env:
             - name: RUST_LOG
               value: {{ .Values.mayastorCP.logLevel }}
+            - name: LLVM_PROFILE_FILE
```
same as above.

blaisedias replied:
fixed https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/c8df674795035432102a90e95b891f9627908aaa

gila commented on control-plane/csi-controller/src/main.rs:
```diff
@@ -97,10 +97,23 @@ pub async fn main() -> Result<(), String> {
+        _evt = signal_int.recv() => {
+                opentelemetry::global::force_flush_tracer_provider();
+            }
+    }
```
really?

blaisedias replied:
cryptic comment, I will assume that we shouldn't invoke `opentelemetry::global::force_flush_tracer_provider()`
Fixed.
https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/fd8a717b34e9e4bc0a8d546f8732f47f0a1d5edc

tiagolobocastro replied:
I think we should keep it actually, it would be useful to flush the traces.

Commit:
fix: remove modifications to helm chart templates

as requested by reviewers

Commit:
fix: review fix, remove open telemetry flush

tiagolobocastro commented on control-plane/msp-operator/src/main.rs:
```diff
@@ -1021,7 +1021,20 @@ async fn main() -> anyhow::Result<()> {
+        _evt = signal_int.recv() => {
+            }
+        _ps = pool_controller(matches) => {
+            _ps?
```
`_` is only used if you don't use the variable

blaisedias replied:
fixed  https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/2c281922104a3665655f0dddce06f56254e18e01

tiagolobocastro commented on control-plane/csi-controller/src/main.rs:
```diff
@@ -97,10 +97,21 @@ pub async fn main() -> Result<(), String> {
+        _srv =  server::CsiServer::run(args.value_of("socket")
+            .expect("CSI socket must be specified")
+            .to_string()) => {
+            _srv?
```
same here with the `_` probably not being needed

gila replied:
I'm missing the reasoning behind this. This code does nothing, so why is it here? Also, why would we need to poll a future that does nothing, all the time? 

I guess it's related to coverage? If so, why is this not backed by a #cfg[(debug_assertions)]? And why dont we make the signal handler actually *do something*?

blaisedias replied:
fixed https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/2c281922104a3665655f0dddce06f56254e18e01

Commit:
fix: review fixes

replace token "cov" with "coverage" to match release and debug
do not use _ prefix for variables which are not ignored

blaisedias commented on nix/pkgs/control-plane/cargo-project.nix:
```diff
@@ -86,4 +90,10 @@ in
       buildType = "debug";
       buildInputs = buildProps.buildInputs;
     });
+  cov = rustNightly.buildRustPackage
```
fixed https://github.com/mayadata-io/mayastor-control-plane/pull/158/commits/2c281922104a3665655f0dddce06f56254e18e01
